### ISSUE TITLE
Manual Implementation of CNN for MNIST Classification

### ISSUE THREAD

**Issue by aidyn-kittrell** [aidyn-kittrell/Deep-Digits-Duel-MLP-vs-CNN#2]
**Goal**: Manually implement the following components (functions):
1. Forward propagation function (including a 2D convolution operation)
2. Backward propagation function (including gradient calculations for convolution kernels and fully connected layer)
3. Train function (for parameter updates of CNN by loss)
4. ReLU activation (for the 1st layer)
5. Softmax activation (for the 2nd layer)
6. Cross-entropy function (for calculating loss)
7. Main function (for training and testing designed by CNN by train dataloader and test dataloader respectively)

**Requirements**:
* Do NOT use any deep learning APIs (e.g., torch.nn, tf.keras, or other pre-built neural network layers)
* Model Architecture, a 2-layer neural network with the following structure:
    1. Convolutional Layer, 1 kernel of size kernel_size x kernel_size (student define the kernel_size), convolution (no padding) + ReLU activation
    2. Fully Connected Layer + 10 neurons with Softmax activation (for 10 classification)
* Implementation Constraints:
    1. Forward Propagation:
        * Input: Batch of MNIST images (shape[batch_size, 28, 28])
        * Output: Prediction probabilities after passing through the two layers
    2. Backward Propagation:
        * Input: Batch of images, ground-truth labels, and predictions from forward propagation
        * Compute gradients for convolution kernels weights and fully connected weights/biases
        * In this pass, compute gradients averaged over the current batch (e.g., divide gradients by the batch size)
        * Use the chain rule to manually derive gradients for both layers
    3. Train Function:
        * Use batch-based training (i.e., process data in batches of size 128 as provided by the dataloader)
        * Input: Training data (images and labels)
        * Iterate over batches of data (maximum 469 iterations per epoch, maximum 5 epochs)
        * Call forward() and backward() in sequence
        * Track and return the training loss
* Additional Specifications:
    1. Use random initialization for weights/biases (e.g., Gaussian distribution for weights, zeros for biases)
    2. Initialize convolution kernels with random values (e.g., Gaussian distribution)
    3. Report/Print test accuracy after training (testing code will be provided)
    4. The should be able to print out the training loop indicator with loss/accuracy logging
    5. Use Python as the programming language
